shared/js/background/newtab-tracker-stats.js

```diff
@@ -0,0 +1,412 @@
+        return {
+            totalCount: this.stats.totalCount,
+            totalPeriod: 'install-time',
+            trackerCompaniesPeriod: 'last-hour',
+            trackerCompanies: stats.results.map(item => {
+                // convert our known key into the 'Other'
+                const displayName = item.key === NewTabTrackerStats.unknownCompanyKey
+                    ? 'Other'
+                    : item.key
+
+                return {
+                    displayName,
+                    count: item.count
+                }
+            })
+        }
+    }
+
+    debounceTimers = {}
+
+    /**
+     * @param {string} name
+     * @param {number} timeout
+     * @param {() => unknown} fn
+     */
+    debounced (name, timeout, fn) {
+        clearTimeout(this.debounceTimers[name])
+        this.debounceTimers[name] = setTimeout(fn, timeout)
+    }
+}
+
+/**
+ * Respond to requests for `tracker-stats.html` - if we determine
+ * that the request was from an allowed origin, re-direct the
+ * request to the web_accessible_resource file 'html/tracker-stats.html'
+ *
+ * @param details
+ */
+export function mv2Redirect () {
+    /**
+     * This listener will redirect the request for tracker-stats.html
+     * on the new tab page to our own HTML file under `web_accessible_resources`
+     */
+    browser.webRequest.onBeforeRequest.addListener((details) => {
+        // NOTE: This part is just for internal testing, it will be removed once the PR is approved
+        if (details.url === chrome.runtime.getURL('html/redirect.html')) {
+            const url = new URL('/chrome_newtab', constants.trackerStats.allowedOrigin)
+            url.searchParams.set('ntp_test', '1')
+            return {
+                redirectUrl: url.toString()
+            }
+        }
+        if (!details.url.endsWith('tracker-stats.html')) {
+            return
+        }
+        if (details.url.startsWith('chrome-extension')) {
+            return
+        }
+        // Only do the redirect if we're being iframed into a known origin
+        if (details.type === 'sub_frame') {
+            const parsed = new URL(details.url)
+            if (parsed.origin === constants.trackerStats.allowedOrigin) {
+                if (parsed.pathname.includes(constants.trackerStats.allowedPathname)) {
+                    return {
+                        redirectUrl: chrome.runtime.getURL(constants.trackerStats.redirectTarget)
+                    }
+                }
+            }
+        }
+        return undefined
+    },
+    {
+        urls: [`${constants.trackerStats.allowedOrigin}/*`],
+        types: ['sub_frame']
+    },
+    ['blocking'])
+}
+
+function mv3Redirect () {
+    const targetUrl = chrome.runtime.getURL(constants.trackerStats.redirectTarget)
+    const incomingUrl = new URL(constants.trackerStats.allowedPathname, constants.trackerStats.allowedOrigin)
+    chrome.declarativeNetRequest.updateDynamicRules({
+        removeRuleIds: [100000000001],
```

Comment:
Is there a better way to do this?

Reply:
CC: @kzar 

can you suggest a way to insert this dynamic rule at runtime, without needing this 'magic number'?